### PR TITLE
feat(@angular/cli): version check the CLI version against npm

### LINKED ISSUE #6592: [TRACKING] New Version Notification
Notify the user when a new version of the CLI or Angular is available.

There is a design that needs review and some more work. @Brocco  you should sync with me to take a look at it.

### PR DESCRIPTION
**Pull request by Brocco**
Fixes #6592

### CODE REVIEW

File: tests/e2e/tests/generate/version-check-not-up-to-date.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import { oneLine } from 'common-tags';
+
+export default function () {
+  return updateJsonFile('node_modules/@angular/cli/package.json', (json) => {
+      json.version = '1.0.0';
+    })
+    .then(() => ng('generate', 'component', 'basic'))
+    .then(({ stdout }) => {
+      if (!stdout.match(/is out of date/)) {
+        throw new Error(oneLine`
```
Comment (hansl) [inline tests/e2e/tests/generate/version-check-not-up-to-date.ts]: This whole throw line fits on one line (less than 100 characters).

File: tests/e2e/tests/generate/version-check-up-to-date.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { ng } from '../../utils/process';
+import { oneLine } from 'common-tags';
+
+export default function () {
+  return ng('generate', 'component', 'basic')
+    .then(({ stdout }) => {
+      if (stdout.match(/is out of date/)) {
+        throw new Error(oneLine`
```
Comment (hansl) [inline tests/e2e/tests/generate/version-check-up-to-date.ts]: This whole throw line fits on one line (less than 100 characters).

File: packages/@angular/cli/update/check.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        resp['dist-tags'].next :
+        resp['dist-tags'].latest;
+      return version;
+    });
+}
+
+function getLocalVersion(rootDir: string, name: string): Promise<string> {
+  const packagePath = path.resolve(rootDir, 'node_modules', name, 'package.json');
+  return new Promise((resolve, reject) => {
+    try {
```
Comment (hansl) [inline packages/@angular/cli/update/check.ts]: Looks like there's no asynchronous workflow in this function. You could just return `Promise.resolve(packageInfo.version)`.

File: packages/@angular/cli/lib/config/schema.json [hunk 4]
```diff
@@ -514,6 +514,28 @@
...
+      "properties": {
+        "lastChecked": {
+          "type": "object",
+          "default": {},
+          "additionalProperties": true
+        },
+        "checkFrequency": {
+          "type": "string",
+          "description": "Defines how often to check for an updated version.",
+          "default": "1d"
```
Comment (hansl) [inline packages/@angular/cli/lib/config/schema.json]: We don't support it, but there's a way for validating strings against regexes in JSON Schema. That would be nice to have here for at least validators and IDEs that support it, and maybe we will soon :)

See http://json-schema.org/latest/json-schema-validation.html#rfc.section.6.8

File: packages/@angular/cli/tasks/version-check.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    if (options.forceCheck || isTimeToCheck) {
+      const rootDir = '';
+      const packageName = '@angular/cli';
+      return newVersionAvailable(rootDir, packageName, options)
+        .then((result: CheckResult) => {
+          const newVersionAvailable = !!result.newVersion;
+          if (newVersionAvailable) {
+            this.ui.writeLine(chalk.yellow(oneLine`Current version (${result.currentVersion})
```
Comment (hansl) [inline packages/@angular/cli/tasks/version-check.ts]: Wording here could be improved with better context and call to action:

```
Your current version of @angular/cli (1.2.3) is out of date. To install the new version, run the following command:
  npm install --save @angular/cli@latest
```

(careful with `--save` vs `--save-dev` depending on where it lives in their package json).

File: packages/@angular/cli/commands/generate.ts [hunk 6]
```diff
@@ -126,6 +126,17 @@ export default Command.extend({
 
     return blueprint.install(blueprintOptions)
       .then(() => {
+        const VersionCheckTask = require('../tasks/version-check').default;
```
Comment (hansl) [inline packages/@angular/cli/commands/generate.ts]: This should happen before the command is run. And we could separate the command output with an empty line.

File: packages/@angular/cli/commands/version-check.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
+const Command = require('../ember-cli/lib/models/command');
```
Comment (hansl) [inline packages/@angular/cli/commands/version-check.ts]: Please remove this command. There's no added value if we have it in generate, build and serve etc.

File: packages/@angular/cli/update/check.ts [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+  const unit = frequency[frequency.length - 1];
+  let factor = 1;
+  switch (unit.toLowerCase()) {
+    case 'h':
+      factor = ONE_HOUR;
+      break;
+    case 'w':
+      factor = ONE_WEEK;
+      break;
+    default:
```
Comment (hansl) [inline packages/@angular/cli/update/check.ts]: `m`?

Comment (Brocco) [inline packages/@angular/cli/update/check.ts]: This is easily achievable by using `30d`/`60d` without having to take into account the number days in a month which is an unnecessary complication for this feature IMHO.

**Comment by Sibiraj-S** [comment]
@Brocco Any Update on this?

**Pull request closed without merging**